src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -1,235 +1,281 @@
-            [toucan.db :as db]))
-
-(defn- constraint->columns [conn constraint-name]
-  (->> ["select column_name from information_schema.constraint_column_usage where constraint_name = ?" constraint-name]
-       (jdbc/query conn)
-       (map :column_name)))
-
-(defn- violates-not-null-constraint [_conn error-message]
-  (let [[match? value column]
-        (re-find #"ERROR:\s+(\w+) value in column \"([^\"]+)\" violates not-null constraint" error-message)]
-    (when match?
-      [{:message (tru "{0} violates not-null constraint" value)
-        :column column}])))
-
-(defn- violates-unique-constraint [conn error-message]
-  (let [[match? constraint _value]
-        (re-find #"ERROR:\s+duplicate key value violates unique constraint \"([^\"]+)\"" error-message)]
-    (when match?
-      (let [columns (constraint->columns conn constraint)]
-        (mapv
-         (fn [column]
-           {:message (tru "violates unique constraint {0}" constraint)
-            :constraint constraint
-            :column column})
-         columns)))))
-
-(defn- update-or-delete-with-fk-constraint [conn error-message]
-  (let [[match? table constraint ref-table _columns _value]
-        (re-find #"ERROR:\s+update or delete on table \"([^\"]+)\" violates foreign key constraint \"([^\"]+)\" on table \"([^\"]+)\"" error-message)]
-    (when match?
-      (let [columns (constraint->columns conn constraint)]
-        (mapv
-         (fn [column]
-           {:message (tru "violates foreign key constraint {0}" constraint)
-            :table table
-            :ref-table ref-table
-            :constraint constraint
-            :column column})
-         columns)))))
+            [toucan.db :as db])
+  (:import
+   (java.sql Connection)))
 
 (defmulti parse-sql-error
   "Parses the raw error message returned after an error in the driver database occurs, and converts it into a sequence
   of maps with a :column and :message key indicating what went wrong."
-  (fn [driver _conn _message] driver))
-
-(defmethod parse-sql-error :default
-  [_driver _conn e]
-  {:message (pr-str e)})
-
-(defmethod parse-sql-error :postgres
-  [_driver conn message]
-  (some #(% conn message)
-        [violates-not-null-constraint
-         violates-unique-constraint
-         update-or-delete-with-fk-constraint]))
+  {:arglists '([driver ^java.sql.Connection connection e]), :added "0.44.0"}
+  driver/dispatch-on-initialized-driver
+  :hierarchy #'driver/hierarchy)
 
 (defn- parse-error
   "Returns errors in a way that indicates which column had the problem. Can be used to highlight errors in forms."
   [driver conn e]
-  (if (not= driver :postgres)
-    (ex-data e)
-    (let [message (ex-message e)]
-      (if-let [errors (and message
-                           (parse-sql-error driver conn message))]
+  (let [default-method? (= (get-method parse-sql-error driver)
+                           (get-method parse-sql-error :default))]
+    (if default-method?
+      (ex-data e)
```

Comment:
It's not a good idea to hardcode the assumption that `:postgres` has an implementation of `parse-sql-error` and no one else does. Instead of that check I tweaked it to check whether the current driver has a method for `parse-sql-error` that's different from the default method.

Reply:
Oh yeah I should mention this but it's idiomatic to write `(if-not (= ...) ...)`, not `(if (not= ...) ...)`. 

https://guide.clojure.style/#if-not